### DISCUSSION TITLE
Passing data to properties that use camel case names in svelte/sveltekit

### DISCUSSION THREAD

**Discussion by OvidijusParsiunas** [lit/lit#4294] [Q&A]
Hi, when working with the Svelte/SvelteKit frameworks - passing data into lit elements generally works fine, however I am experiencing issues when attempting to pass data from state into my element's properties that use camel cases for their names.

Example:

```
<script>
    const localObject = {...}
</script>
<main>
  
<custom-lit-element
    property={content} // this property passes the object
    camelProperty={content} // this property passes null
></custom-lit-element>
</main>
```

As a workaround I can use the following approach, however this does not work well for passing functions as they don't keep state to the rest of the component:
```
camelProperty={JSON.stringify(content)} 
```

What is happening here is that Svelte is automatically lower casing the property names and because they no longer match the expected property names - they are assumed to be attributes (hence the stringified approach works). This is also evident by the following line in [their codebase](https://github.com/sveltejs/svelte/blob/6f508a011bc8051ab9f82cafa97c5292a4453b92/packages/svelte/src/runtime/internal/dom.js#L421):

```
else if (node.getAttribute(attribute) !== value)node.setAttribute(attribute, value);
```

I was wondering if there is any way to pass data into properties with camel cases using Lit Element.

Here is a [live example](https://codesandbox.io/s/lit-svelte-property-problem-mxyqnx?file=/App.svelte) where passing values into a property called `frameComponentsStyles` can only be done when it is stringified. For reference [`frameComponentsStyles`](https://github.com/OvidijusParsiunas/active-table/blob/e5b81cf62eb1e5f8b7ad8208a4bc13abaee07ddd/component/src/activeTable.ts#L206C43-L206C43) is a simple property that is set up using the following code where [FrameComponentsStyles](https://github.com/OvidijusParsiunas/active-table/blob/e5b81cf62eb1e5f8b7ad8208a4bc13abaee07ddd/component/src/types/frameComponents.ts#L17) type is an object:

```
@property({type: Object})
frameComponentsStyles: FrameComponentsStyles = {};
```

Thankyou for your feedback!

**Comment by justinfagnani** [comment]
I thought this would have been covered by custom-elements-everywhere.com, but apparently all the property binding tests use lower-case-only names: https://github.com/webcomponents/custom-elements-everywhere/blob/main/libraries/svelte/src/components/ComponentWithProperties.svelte

Even so, looking at the Svelte code I'm not sure that this is an issue with camelCase property names. The `attr()` function you linked to never sets properties at all, and we know that Svelte sets properties in some cases. The function where `attr()` is called from appears to be `set_attributes()`, which checks for the presence of the property on the prototype and sets if it exists.

I can see a bug there where the property logic only kicks in for [own properties](https://github.com/sveltejs/svelte/blob/6f508a011bc8051ab9f82cafa97c5292a4453b92/packages/svelte/src/runtime/internal/dom.js#L439) of the prototype though, so this logic won't work for inherited accessors. That could be done to avoid setting properties for accessors on HTMLElement, but it breaks inheritance.

But don't see this inheritance pattern in ActiveTable - `frameComponentsStyles` is defined right on the class.

The only other things I could think of are 
 1. That there's some kind of proxy in front of ActiveTable that makes it not the direct prototype, but that doesn't appear to be the case
 2. There's some timing issue where `<active-table>` isn't defined when the property/attribute is set so the prototype check fails and Svelte sets an attribute. I didn't see a quick way to check for that in that sandbox. I think we'd need to use a bespoke custom element to diagnose that.

**Reply by OvidijusParsiunas** [reply]
Thankyou for the response @justinfagnani.

I have experienced this problem before on another web component that I built using vanilla js/ts (not Lit) and the way I got around it was by automatically creating lower cased properties for all camel cased properties which would update the camel cased ones if they were set with new values. It works like magic, but I am not sure how possible this is in Lit element.
For reference the code for it is [here](https://github.com/OvidijusParsiunas/deep-chat/blob/1f8d22f9a7d2ade0a7dd1c4d6519ceb6a8705636/component/src/utils/webComponent/internalHTML.ts#L29).

In regards to your response, I apologise but I don't think understand what the next steps should be.
I have created another sandbox with a very basic Lit element web component and the result appears to be the same, meaning that the problem is not isolated to Active Table:
[Live example](https://codesandbox.io/s/lit-svelte-property-problem-2-m3csg5?file=/App.svelte)

Let me know if theres anything I can assist with, thankyou!

**Reply by justinfagnani** [reply]
ok, it indeed is a Svelte camelCase bug.

I reduced the reproduction to not include Lit, which I think will help get traction on the Svelte side: https://codesandbox.io/s/svelte-web-component-property-problem-2-forked-kl25tw

The next step would be to file a bug on Svelte. I hope they're receptive to it.

**Reply by OvidijusParsiunas** [reply]
Thankyou for the reply! I am more than happy to raise a new issue.

Just to note, I did some digging around and found a couple of existing recent issues relating to this problem in the [Svelte](https://github.com/sveltejs/svelte) repo:

- The [first one](https://github.com/sveltejs/svelte/issues/5184) was closed with a comment stating:
_"This behavior is correct and in line with how other custom element libraries like Lit do it."_

- And [another one](https://github.com/sveltejs/svelte/pull/7638) closed with:
_lowercase to camelCase conversion will be done by default ([as inspired by Lit](https://lit.dev/docs/components/properties/#observed-attributes) per https://github.com/sveltejs/svelte/pull/8457)._
In other words specifying the attribute name on the property should fix it. I have tested this using their very latest `4.2.1` version and the problem is still there.

It seems that svelte devs/users are expecting this to work correctly in Lit or perhaps I may be misunderstanding this and they are noting that this problem is to be expected...? Just wanted to share this as it was linked to the problem in this discussion.

Finally, my apologies again, when you commented _"The next step would be to file a bug on Svelte"_ is it ok if I do it, or did you imply that you would. As I mentioned I am very happy to do it but don't want to create one if you were going to. Thanks again!

**Reply by sorvell** [reply]
There might be some confusion here, and I hope I'm not adding to it...

Lit correctly maps attributes which are not case sensitive to properties which are. It does this in `attributeChangedCallback` by checking in the if the given attribute has a corresponding reactive property.

That's not what's happening here since we want/expect Svelte  to set the property directly and not the attribute, to avoid the need to/difficulty of passing data with strings.

To do this correctly, Svelte must preserve case when it checks whether or not to set a property or attribute on a custom element.

**Reply by OvidijusParsiunas** [reply]
Hey @sorvell, thankyou for the message! Your statement makes sense. Svelte is not preserving property name casing which in turn makes us fallback to using attributes. This is problematic as objects then need to be serialized to strings.

The last [comment](https://github.com/sveltejs/svelte/pull/7638#issuecomment-1569476944) from an [issue](https://github.com/sveltejs/svelte/pull/7638) I have mentioned above states:
_"The kebab-case to camelCase conversion will not happen automatically but must instead be explicitly mapped"_

Which forces devs to look for workarounds in their _framework agnostic_ web component implementations. I don't think this should be the way.

**Reply by justinfagnani** [reply]
To be clear, this isn't about attributes, it's about properties. Svelte needs to be able to set properties with valid names, and it apparently can't.

I'm updating custom-elements-everywhere to add a camelCase property test: https://github.com/webcomponents/custom-elements-everywhere/pull/2239

A few frameworks don't pass it. Hopefully that helps motivate a new Svelte issue.

**Reply by OvidijusParsiunas** [reply]
Much appreciated @justinfagnani, I'll hold off for that PR to be closed before creating an issue just to reference it in the initial description.

**Reply by OvidijusParsiunas** [reply]
I have raised a new [issue](https://github.com/sveltejs/svelte/issues/9325) and will update this thread if there are any _significant_ updates.

**Reply by justinfagnani** [reply]
Thanks for the new issue! That looks like one I link from custom-elements-everywhere. You can see the new failing camelCase test results for Svelte here: https://custom-elements-everywhere.com/libraries/svelte/results/results.html

I also have a PR adding a test for setting inherited properties: https://github.com/webcomponents/custom-elements-everywhere/pull/2263

Svelte will need a new issue for that too. Once that PR is merged we can link to the results page for it too.

**Reply by justinfagnani** [reply]
If you can link to that results page in the new issue that might be helpful.

**Reply by OvidijusParsiunas** [reply]
Hey @justinfagnani, I am more than happy to raise another issue, but am wondering is the property inheritance problem related to the camel case problem as inherited properties seem to work fine for me:

[Live CodeSandbox](https://codesandbox.io/s/svelte-web-component-property-inheritance-problem-wygzsh?file=/inheritance-element.js)

I could be misunderstanding this. Let me know if there are any issues with the sandbox, thanks!

**Reply by OvidijusParsiunas** [reply]
If it is still to do with camel casing I should perhaps update the existing issue instead to help them take inheritance into consideration when coming up with a fix.

**Reply by dummdidumm** [reply]
Thanks for bringing this to our attention - the new Svelte version fixing this was released, which hopefully brings Svelte up to 100% tests passing again

**Reply by OvidijusParsiunas** [reply]
Thanks again for your work @dummdidumm!

**Reply by OvidijusParsiunas** [reply]
The camelCase issue that I was facing has now been resolved.

I know that there is a PR that includes tests for inherited properties and I have raised [another PR](https://github.com/webcomponents/custom-elements-everywhere/pull/2274) to bump up the tested Svelte version, so I'll wait for @justinfagnani's word before closing this discussion.

I really appreciate all the effort you've put into this folks!

**Reply by OvidijusParsiunas** [reply]
Interestingly the build failed for that PR failed, I'll wait for further comments before taking any action.

**Reply by OvidijusParsiunas** [reply]
Hey folks, there haven't been any updates on this for a while and the core objective of this discussion has already been accomplished, therefore I am going to be closing this discussion.

 If there is anything else that needs to be done, feel free to comment or re-open the discussion as needed.

Once again, thank you for all the work you've put into this, I really appreciate it!

**Reply by justinfagnani** [reply]
Thanks for raising the issue @OvidijusParsiunas !
